src/middlewared/middlewared/gdb.py

```diff
@@ -0,0 +1,190 @@
+import argparse
+import logging
+import libzfs
+import os
+import re
+import requests
+import subprocess
+import tempfile
+import textwrap
+import tqdm
+
+
+class MiddlewareGDB(object):
+
+    def __init__(self):
+        self.logger = logging.getLogger('middlewaregdb')
+
+    def get_dataset(self, name=None):
+        zfs = libzfs.ZFS()
+        if name is None:
+            for pool in zfs.pools:
+                if pool.name == 'freenas-boot':
+                    continue
+                return pool.root_dataset
+            raise RuntimeError('no dataset found')
+        else:
+            return zfs.get_dataset(name)
+
+    def _get_version(self):
+        with open('/etc/version', 'r') as f:
+            version = f.read().split()[0]
+
+        product, version = version.split('-', 1)
+        major = version.split('-')[0].split('.')[0]
+
+        return {
+            'major': major,
+            'product': product,
+            'version': version,
+        }
+
+    def debug_symbols_url(self):
+        version = self._get_version()
+        if version['product'] == 'FreeNAS':
+            base_url = f'https://download.freenas.org/{version["major"]}'
+            if 'MASTER' in version['version']:
+                verdir = f'MASTER/{version["version"].split("-")[-1]}'
+            else:
+                verdir = version['version']
+            dirurl = f'{base_url}/{verdir}/x64/'
+            req = requests.get(dirurl)
+            if req.status_code != 200:
+                raise RuntimeError('debug symbols url could not be found, provide the url')
+            reg = re.search(r'href="(?P<debug>.+?\.debug\.txz)"', req.text, re.M)
+            if reg is None:
+                raise RuntimeError(f'debug symbols file not found in {dirurl}')
+            return f'{dirurl}{reg.group("debug")}'
+        else:
+            raise RuntimeError(f'product {version["product"]} not supported')
+
+    def _download_debug(self, dataset, url):
+        filename = url.rsplit('/', 1)[-1]
+        path = f'{dataset.mountpoint}/{filename}'
+
+        if os.path.exists(path):
+            first_byte = os.path.getsize(path)
+        else:
+            first_byte = 0
+
+        headers = {}
+        r = requests.head(url)
+        total_size = int(r.headers.get('content-length', 0))
+        if total_size:
+            headers['Range'] = f'bytes={first_byte}-{total_size}'
+
+        if first_byte and first_byte == total_size:
+            return path
+
+        if first_byte > 0 and total_size == first_byte:
```

Comment:
Two duplicate conditions in a row

We can just call `subprocess.check_call(["wget", "-c", ...])` here instead of inventing the wheel

Reply:
I dont believe this is reinventing the wheel. I dislike calling subprocess and this we can control whatever happens in case of an error and also show a progress bar.

I will happily change it if there is any big disadvantage I am missing

Reply:
I don't see any disadvantages in keeping current solution, let's keep it. I just like `subprocess + wget` approach. It handles (and retries) all errors that can occur (e.g. you did not add `timeout`, this is not important for a cli tool, but can be useful somewhere else), throws exception if errors are fatal) and with `check_call` progress bar is displayed properly.